plugins/teststeps/teststeps.go

```diff
@@ -29,94 +30,66 @@ type PerTargetFunc func(cancel, pause <-chan struct{}, target *target.Target) er
 // each target. The implementation of the per-target function is responsible for
 // handling internal cancellation and pausing.
 func ForEachTarget(pluginName string, cancel, pause <-chan struct{}, ch test.TestStepChannels, f PerTargetFunc) error {
-	type tgtErr struct {
-		target *target.Target
-		err    error
+	// Temporary solution
+	ctx, cancelCtx := combineChannelsToContext(cancel, pause)
+	defer cancelCtx()
+
+	reportTarget := func(t *target.Target, err error) {
+		if err != nil {
+			log.Errorf("%s: ForEachTarget: failed to apply test step function on target %s: %v", pluginName, t, err)
+			select {
+			case ch.Err <- cerrors.TargetError{Target: t, Err: err}:
+			case <-ctx.Done():
+				log.Debugf("%s: ForEachTarget: received cancellation/pause signal while reporting error", pluginName)
+			}
+		} else {
+			log.Debugf("%s: ForEachTarget: target %s completed successfully", pluginName, t)
+			select {
+			case ch.Out <- t:
+			case <-ctx.Done():
+				log.Debugf("%s: ForEachTarget: received cancellation/pause signal while reporting error", pluginName)
+			}
+		}
 	}
-	tgtErrCh := make(chan tgtErr)
-	tgtInFlight := int32(0)
-	noMoreTargetsCh := make(chan struct{})
 
-	go func(tgtInFlight *int32) {
-		defer func() {
-			log.Debugf("%s: ForEachTarget: exiting incoming loop, targets inflight %d", pluginName, atomic.LoadInt32(tgtInFlight))
-		}()
+	var wg sync.WaitGroup
+	func() {
 		for {
 			select {
 			case tgt := <-ch.In:
 				log.Debugf("%s: ForEachTarget: received target %s", pluginName, tgt)
 				if tgt == nil {
 					log.Debugf("%s: ForEachTarget: all targets have been received", pluginName)
-					noMoreTargetsCh <- struct{}{}
 					return
 				}
 
+				wg.Add(1)
 				go func() {
-					tgtErrCh <- tgtErr{target: tgt, err: f(cancel, pause, tgt)}
+					defer wg.Done()
+
+					err := f(cancel, pause, tgt)
+					reportTarget(tgt, err)
 				}()
-				atomic.AddInt32(tgtInFlight, 1)
-			case <-cancel:
-				log.Debugf("%s: ForEachTarget: incoming loop canceled", pluginName)
-				return
-			case <-pause:
-				log.Debugf("%s: ForEachTarget: incoming loop paused", pluginName)
+			case <-ctx.Done():
+				log.Debugf("launching new targets has been cancelled")
```

Comment:
Logs are really hard to read without the plugin
```suggestion
				log.Debugf("%s: ForEachTarget: incoming loop canceled", pluginName)
```